### PR TITLE
Doc: Add docs for running PyInstaller with Python optimizations

### LINKED ISSUE #2447: Document running PyInstaller with `-O`
Document how to make PyInstaller creating optimized (`-O`, `-OO`) code, see #2442.

You can simply run `python -O /path/to/pyinstaller/` or `python -O -m PyInstaller`.

I have to admit that this is not obvious since pyinstaller is installed via pip, but it still works. I'd not spend too much effort on this as `.pyo` files are bound to be removed soon (no source at hand ATM).

### PR DESCRIPTION
**Pull request by nejch**
Closes #2447.

### CODE REVIEW

File: doc/usage.rst [hunk 1]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
         myscript.spec
 
 
+Running PyInstaller with Python optimizations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (htgoebel) [inline doc/usage.rst]: The underlining must be at least as long (in number of characters) as the heading.

File: doc/usage.rst [hunk 2]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
...
+
+.. Note::
+
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
+	this option. Using ``-O``, ``__debug__`` is set to ``False`` and
```
Comment (htgoebel) [inline doc/usage.rst]: "When using …"

File: doc/usage.rst [hunk 3]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
...
+	``assert`` statements are not executed. The ``-OO`` option
```
Comment (htgoebel) [inline doc/usage.rst]: assert statements are not only "not executed", but completely removed from the byte-code.

File: doc/usage.rst [hunk 4]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
         myscript.spec
 
 
+Running PyInstaller with Python optimizations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. Note::
+
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
```
Comment (htgoebel) [inline doc/usage.rst]: * Well, this is not an "option" for pyinstaller, so I suggest you using different wording.
* I'd start the sentence with the "When using this …"
* "be aware of" sound like: you should know there is such a think. But one needs no know what this does.
* ``PYTHONOPTIMIZE`` is the name of the env-var. Maybe better just "Python Bytecode Optimization mechanism"

Suggestion:
> When using this feature/option/thing, you should be aware of how the the ``PYTHONOPTIMIZE`` mechanism works.

File: doc/usage.rst [hunk 5]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
...
+	``assert`` statements are not executed. The ``-OO`` option
+	additionally omits docstrings. If your code relies on any of these
+	features, your program may break or have unexpected behavior.
+
+PyInstaller can be run with Python optimization options (``-O`` or ``-OO``)
+by executing it as a Python module, rather than using the ``pyinstaller`` command::
+
+    python -O -m PyInstaller myscript.py
+
+You can use any PyInstaller options that are otherwise available with the ``pyinstaller`` command. For example::
```
Comment (htgoebel) [inline doc/usage.rst]: long line.

File: doc/usage.rst [hunk 6]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
...
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
+	this option. Using ``-O``, ``__debug__`` is set to ``False`` and
+	``assert`` statements are not executed. The ``-OO`` option
+	additionally omits docstrings. If your code relies on any of these
+	features, your program may break or have unexpected behavior.
+
+PyInstaller can be run with Python optimization options (``-O`` or ``-OO``)
+by executing it as a Python module, rather than using the ``pyinstaller`` command::
+
+    python -O -m PyInstaller myscript.py
```
Comment (htgoebel) [inline doc/usage.rst]: Want to add these, too?
```
python -OO -m PyInstaller myscript.py
PYTHONOPTIMIZE=1 pyinstaller myscript.py
```

File: doc/usage.rst [hunk 7]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
...
+
+You can use any PyInstaller options that are otherwise available with the ``pyinstaller`` command. For example::
+
+    python -O -m PyInstaller --onefile myscript.py
+
+Alternatively, you can also use the path to pyinstaller::
+
+    python -O /path/to/pyinstaller/ myscript.py
```
Comment (htgoebel) [inline doc/usage.rst]: I assume there should be no trailing slash. Maybe use `/path/to/bin/pyinstaller`?

File: doc/usage.rst [hunk 8]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
         myscript.spec
 
 
+Running PyInstaller with Python optimizations
```
Comment (htgoebel) [inline doc/usage.rst]: PyInstaller should be written `|PyInstaller|` anywhere.

File: doc/usage.rst [hunk 9]
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. Note::
+
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
+	this option. Using ``-O``, ``__debug__`` is set to ``False`` and
+	``assert`` statements are not executed. The ``-OO`` option
+	additionally omits docstrings. If your code relies on any of these
+	features, your program may break or have unexpected behavior.
```
Comment (htgoebel) [inline doc/usage.rst]: We should add something like "Using this feature/option/thing effects not only your main script but *all* modules compiled by |PyInstaller|"

**Comment by nejch** [comment]
Thanks for the feedback, sorry I wasn't aware of the `[skip ci]` tag. I've added the changes requested and built it with sphinx to test. I changed the option wording to 'feature' (and 'flag' for `-O`) to avoid confusion with PyInstaller options.

The snippets might a bit wordy now as I added a separate Windows line for running with `PYTHONOPTIMIZE=1`, see if it's ok though. I'll push the changes in a minute.

File: doc/usage.rst [hunk 10]
```diff
@@ -89,6 +89,49 @@ Or in Windows, use the little-known BAT file line continuation::
...
+
+.. Note::
+
+    When using this feature, you should be aware of how the Python bytecode
+    optimization mechanism works. When using ``-O``, ``__debug__`` is set
+    to ``False`` and ``assert`` statements are removed from the bytecode.
+    The ``-OO`` flag additionally removes docstrings.
+
+    Using this feature affects not only your main script, but *all* modules
+    compiled by |PyInstaller|. If your code (or any module imported by your
```
Comment (htgoebel) [inline doc/usage.rst]: This should be "included by PyInstaller", since PyInstaller will not compile the source if there already is a .pyo file.

**Comment by htgoebel** [comment]
Thanks for the update. Sorry for not applying it earlier.
